### PR TITLE
[MLv2] Support Segment Filters

### PR DESCRIPTION
**Pull request by iethree**
Closes https://github.com/metabase/metabase/issues/34495

> **Note** 
> Disregard the clojure changes, I pulled them in from https://github.com/metabase/metabase/pull/3454

### Description

Supports segment filters by extending the columnPicker to support segments if there is an `onSegmentSelect` function provided

Screen Shot 2023-10-12 at 4 23 42 PM

### CODE REVIEW

File: frontend/src/metabase/common/components/QueryColumnPicker/QueryColumnPicker.tsx [hunk 1]
```diff
@@ -48,26 +59,42 @@ export function QueryColumnPicker({
...
+        const includeSegments = !!onSelectSegment && !groupInfo.isFromJoin;
+
+        const segmentItems = includeSegments
+          ? Lib.availableSegments(query).map(
+              (segment: Lib.SegmentMetadata) => ({
+                ...Lib.displayInfo(query, stageIndex, segment),
+                segment,
+              }),
+            )
+          : [];
```
Comment (iethree) [inline frontend/src/metabase/common/components/QueryColumnPicker/QueryColumnPicker.tsx:86]: I don't *love* overloading the columnPicker to be a column-or-segment picker, but it seemed insane to create a separate picker that duplicated all of the column picker logic. 

I wonder if it's worth reworking the columnPicker to be more generic? I also fear that might be grossly over-engineering it.

File: frontend/src/metabase/common/components/QueryColumnPicker/QueryColumnPicker.tsx [hunk 2]
```diff
@@ -48,26 +59,42 @@ export function QueryColumnPicker({
...
-          column,
-        }));
+        const columnItems = Lib.getColumnsFromColumnGroup(group).map(
+          column => ({
+            ...Lib.displayInfo(query, stageIndex, column),
+            column,
+          }),
+        );
+
+        const includeSegments = !!onSelectSegment && !groupInfo.isFromJoin;
```
Comment (iethree) [inline frontend/src/metabase/common/components/QueryColumnPicker/QueryColumnPicker.tsx:77]: I [posted in slack](https://metaboat.slack.com/archives/C04CYTEL9N2/p1697148236490379) about this, but currently both MLv1 and MLv2 only support segments for non-joined tables. This seems like a bug to me, but maybe not one worth fixing right now.

**Review by ranquild — COMMENTED** [review]
I'd rather not extend `QueryColumnPicker` but create a separate component for filters only. Even the name is misleading - we no longer pick columns but also segments. Also most of the functionality in the `QueryColumnPicker` is not needed for filters - e.g. binning or temporal bucketing. 

For aggregations we have `AggregationPicker` which renders a combined list of operators and metrics in once place. I think we need to do the same here - e.g. add our own component.

@kulyk WDYT?

**Comment by kulyk** [comment]
> For aggregations we have `AggregationPicker` which renders a combined list of operators and metrics in once place. I think we need to do the same here - e.g. add our own component.

`AggregationPicker` is a bit different, it doesn't mix anything into a columns list, we're only [rendering `QueryColumnPicker` when selecting a column for a selected aggregation operator](https://github.com/metabase/metabase/blob/c2209191610666afe26e6877083aa7961b75e931/frontend/src/metabase/common/components/AggregationPicker/AggregationPicker.tsx#L239). Here it's a bit more mixed up (columns + segments + custom expression button).

> I'd rather not extend `QueryColumnPicker` but create a separate component for filters only. Even the name is misleading - we no longer pick columns but also segments. Also most of the functionality in the `QueryColumnPicker` is not needed for filters - e.g. binning or temporal bucketing.

Yeah, I'm not a big fan of how we're extending the `QueryColumnPicker` both here and in #34603
But I don't like having two _very_ similar column pickers either :( OTOH, it's not like `QueryColumnPicker` has a lot of code. Most of it is needed for bucketing, and we don't need that for filters as you pointed out. And the rest of the component is basically a wrapper around `AccordionList`

**Comment by kulyk** [comment]
Looking more closely, I agree with @ranquild that we should better make another component. It's not great we need to duplicate some code, but it's not much. I opened [another PR adding custom expressions](https://github.com/metabase/metabase/pull/34656) and it also migrates filters off the `QueryColumnPicker`. The fact we're building `AccordionList's` sections right there should make adding segments more straightforward

File: frontend/src/metabase-lib/metadata.ts [hunk 3]
```diff
@@ -29,11 +31,11 @@ import type {
...
   OrderByClause,
   OrderByClauseDisplayInfo,
+  Query,
+  SegmentMetadata,
+  SegmentDisplayInfo,
   TableDisplayInfo,
   TableMetadata,
-  Query,
-  FilterOperator,
-  FilterOperatorDisplayInfo,
```
Comment (kulyk) [inline frontend/src/metabase-lib/metadata.ts:36]: ❤️

File: frontend/src/metabase-lib/types.ts [hunk 4]
```diff
@@ -141,6 +141,14 @@ export type ColumnDisplayInfo = {
   selected?: boolean; // used in aggregation and field clauses
 };
 
+export type SegmentDisplayInfo = {
+  name: string;
+  displayName: string;
+  longDisplayName: string;
+  description: string;
+  effectiveType: string;
+};
```
Comment (kulyk) [inline frontend/src/metabase-lib/types.ts:150]: Do we really need the `effectiveType` here? I'd rather prefer if the FE didn't know about it at all

Comment (iethree) [reply frontend/src/metabase-lib/types.ts:150]: I just typed it because MLv2 gives it to us. 🤷

Comment (kulyk) [reply frontend/src/metabase-lib/types.ts:150]: I think we can skip it 👍

File: frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  getColumnGroupIcon,
+  getColumnGroupName,
+} from "metabase/common/utils/column-groups";
+
+import { Icon } from "metabase/core/components/Icon";
+import type { IconName } from "metabase/core/components/Icon";
+
+import * as Lib from "metabase-lib";
+
+import { StyledAccordionList } from "../FilterPicker.styled";
```
Comment (kulyk) [inline frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx:15]: Can we please move this to `FilterColumnPicker.styled`? Other components won't need the `StyledAccordionList`

File: frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+};
+
+export type SegmentListItem = Lib.SegmentDisplayInfo & {
+  segment: Lib.SegmentMetadata;
+};
+
+const isSegmentListItem = (
+  item: ColumnListItem | SegmentListItem,
+): item is SegmentListItem => {
+  return (item as SegmentListItem).segment !== undefined;
```
Comment (kulyk) [inline frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx:50]: How about something like that?

```suggestion
return (item as SegmentListItem).segment != null;
```

With the current check, `isSegmentListItem` would return `true` if `segment` is `null`

File: frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    const sections = columnGroups.map(group => {
+      const groupInfo = Lib.displayInfo(query, stageIndex, group);
+
+      const columnItems = Lib.getColumnsFromColumnGroup(group).map(column => ({
+        ...Lib.displayInfo(query, stageIndex, column),
+        column,
+      }));
+
+      const includeSegments = !groupInfo.isFromJoin;
```
Comment (kulyk) [inline frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx:76]:
```suggestion
      const includeSegments = groupInfo.isSourceTable;
```

With `!groupInfo.isFromJoin` we'd also include segments in implicitly joinable table groups

File: frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+      const segmentItems = includeSegments
+        ? Lib.availableSegments(query).map((segment: Lib.SegmentMetadata) => ({
+            ...Lib.displayInfo(query, stageIndex, segment),
+            segment,
+          }))
+        : [];
```
Comment (kulyk) [inline frontend/src/metabase/common/components/FilterPicker/FilterColumnPicker/FilterColumnPicker.tsx:83]: We shouldn't need to type the `segment` argument as long as `Lib.availableSegments` has a return type

**Review by kulyk — CHANGES_REQUESTED** [review]
@iethree thanks for addressing the comments! Inlined a few more suggestions

Also, it seems like we'd need to make the `FilterPicker` slightly less column-centric. When clicking a filter clause that uses a segment, the picker opens the expression editor by default:

demo

File: frontend/src/metabase/query_builder/components/notebook/steps/FilterStep/FilterStep.tsx [hunk 9]
```diff
@@ -84,6 +84,7 @@ export function FilterStep({
             onClose={onClose}
           />
         )}
+        withLegacyPopover
```
Comment (iethree) [inline frontend/src/metabase/query_builder/components/notebook/steps/FilterStep/FilterStep.tsx:87]: without this (which I copied from AggregationStep) the popover was aggressively closing whenever you clicked an expression suggestion

Comment (iethree) [reply frontend/src/metabase/query_builder/components/notebook/steps/FilterStep/FilterStep.tsx]: ok, removed that commit

**Review by kulyk — APPROVED** [review]
LGTM!

**Pull request merged**